aria/orchestrator/execution_plugin/ctx_proxy/server.py

```diff
@@ -170,101 +178,59 @@ def _process_ctx_request(ctx, args): # pylint: disable=too-many-branches,too-man
-                # Don't count initial args fulfilled by the partial
-                spec_args = argspec.args[len(current.args):]
-                # Don't count keyword args fulfilled by the partial
-                spec_args = tuple(a for a in spec_args if a not in current.keywords)
-            else:
-                argspec = inspect.getargspec(current)
-                spec_args = argspec.args
-
-            callable_kwargs = {}
-            if isinstance(arg, dict):
-                # If the first arg is a dict, treat it as our kwargs
-                # TODO: what if the first argument really needs to be a dict?
-                callable_kwargs = arg
-                index += 1
-
-            if argspec.varargs is not None:
-                # Gobble the rest of the args
-                callable_args = args[index:]
-            else:
-                # Take only what we need
-                spec_args = tuple(a for a in spec_args if a not in callable_kwargs)
-                spec_args_count = len(spec_args)
-                if inspect.ismethod(current):
-                    # Don't count "self" argument
-                    spec_args_count -= 1
-                callable_args = args[index:index + spec_args_count]
-                # Note: we might actually have fewer args than the args_count, but that could be OK
-                # if the user expects subsequent args to have default values
-
-            args_count = len(callable_args)
-            if args_count > 1:
-                index += args_count - 1
-
-            current = current(*callable_args, **callable_kwargs)
-
-        else:
-            raise RuntimeError('`{0}` cannot be processed in {1}'.format(arg, args))
-
-        index += 1
-
-    if callable(current):
-        current = current()
+    # Parse all arguments
+    while len(args) > 0:
+        obj, args = _parse_argument(obj, args, modifying)
 
     if modifying:
-        if hasattr(current, '__setitem__'):
-            # Modify dict value
-            current[modifying_key] = modifying_value
-        else:
+        if hasattr(obj, '__setitem__'):
+            # Modify item value (dict, list, and similar)
+            if isinstance(obj, (list, tuple)):
+                modifying_key = int(modifying_key)
+            obj[modifying_key] = modifying_value
+        elif hasattr(obj, modifying_key):
             # Modify object attribute
-            setattr(current, modifying_key, modifying_value)
-
-    return current
+            setattr(obj, modifying_key, modifying_value)
+        else:
+            raise ProcessingError('Cannot modify `{0}` of `{1!r}`'
+                                  .format(modifying_key, obj))
 
+    return obj
 
-def _desugar_attr(obj, attr):
-    if not isinstance(attr, basestring):
-        return None
-    if hasattr(obj, attr):
-        return attr
-    attr = attr.replace('-', '_')
-    if hasattr(obj, attr):
-        return attr
-    return None
 
+def _parse_argument(obj, args, modifying):
```

Comment:
i would consider splitting this function into two different functions (since the only usage for the entire args is being made in case of a function call